tower-http/src/cors.rs

```diff
@@ -0,0 +1,627 @@
+        self.update_layer(|layer| layer.allow_credentials(allow_credentials))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Headers`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_headers`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers
+    pub fn allow_headers<I>(self, headers: I) -> Self
+    where
+        I: IntoIterator<Item = HeaderValue>,
+    {
+        self.update_layer(|layer| layer.allow_headers(headers))
+    }
+
+    /// Set the value of the [`Access-Control-Max-Age`][mdn] header.
+    ///
+    /// See [`CorsLayer::max_age`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+    pub fn max_age(self, max_age: Duration) -> Self {
+        self.update_layer(|layer| layer.max_age(max_age))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Methods`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_methods`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
+    pub fn allow_methods<T>(self, methods: T) -> Self
+    where
+        T: Into<AnyOr<Vec<Method>>>,
+    {
+        self.update_layer(|layer| layer.allow_methods(methods))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Origin`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_origin`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+    pub fn allow_origin<T>(self, origin: T) -> Self
+    where
+        T: Into<AnyOr<Origin>>,
+    {
+        self.update_layer(|layer| layer.allow_origin(origin))
+    }
+
+    /// Set the value of the [`Access-Control-Expose-Headers`][mdn] header.
+    ///
+    /// See [`CorsLayer::expose_headers`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+    pub fn expose_headers<I>(self, headers: I) -> Self
+    where
+        I: Into<AnyOr<Vec<HeaderName>>>,
+    {
+        self.update_layer(|layer| layer.expose_headers(headers))
+    }
+
+    fn update_layer<F>(mut self, f: F) -> Self
+    where
+        F: FnOnce(CorsLayer) -> CorsLayer,
+    {
+        self.layer = f(self.layer);
+        self
+    }
+
+    fn is_valid_origin(&self, origin: &HeaderValue) -> bool {
+        match &self.layer.allow_origin.0 {
+            AnyOrInner::Any => true,
+            AnyOrInner::Value(allow_origin) => match &allow_origin.0 {
+                OriginInner::Exact(s) => s == origin,
+                OriginInner::List(list) => list.contains(origin),
+            },
+        }
+    }
+
+    fn build_preflight_response<B>(&self, origin: HeaderValue) -> Response<B>
+    where
+        B: Default,
+    {
+        let mut response = Response::new(B::default());
```

Comment:
I think preflight should have no body which I don't think is guaranteed by `::default()`. I doubt any of the browser implementations depend on that, so, it's probably more of an ideal than anything. 

Still is `impl Default` for `Body` typical in `tower` and elsewhere? would it be more clear to have a different trait `EmptyBody` or requiring this from the builder..

Reply:
All bodies in http-body, axum, and tower-http are empty in their `Default` impl. While I see your point I think its safe for now. Can fix if someone reports issues with it.